### ISSUE TITLE
WHAM keeps failing with insufficient overlap

### ISSUE THREAD

**Issue by jhprinz** [openpathsampling/openpathsampling#549]
I keep getting errors from WHAM about

```
RuntimeError ("Insufficient overlap to combine histograms.
```

not always but quite often.

I think this is due to too small bin sizes, is that correct? If so, could we just make the bins a little larger in the examples?

**Comment by dwhswenson** [comment]
> I think this is due to too small bin sizes, is that correct?

No, it is due to insufficient sampling. This error means that there is no overlap in the histograms. Changing the bin sizes will only fix it if you make the bins larger than the space between interfaces.

**Comment by dwhswenson** [comment]
One thing that might be worth changing: are we still using the *square* of the distance as an order parameter? I originally set it up that way, thinking that we'd save computational cost by skipping the `sqrt`. However, our overhead is so much more than that `sqrt`, so it might be worth including it in the order parameter.

I don't think it will fix this problem (you still won't have overlap between interfaces), but it would probably make more intuitive sense to work in a linear space instead a quadratic one.

**Comment by jhprinz** [comment]
I changed the binsize in one example from `0.02` to `0.05` and then it worked. Not sure why then.
Can that make sense?

**Comment by dwhswenson** [comment]
Maybe, although partly by luck and partly by doing bad science. The interfaces are at 0.00, 0.09, 0.16, 0.25. And your bins edges at 0.0, 0.05, 0.10, etc. So probably you're saying that a result of 0.06 crosses the interface at 0.09. In other words, your bins are too wide to be meaningful.

Really, we should switch to a linear order parameter. Then we need actually sufficient sampling.

**Comment by jhprinz** [comment]
Got it. Good point. Can you change that?

**Comment by dwhswenson** [comment]
Yeah -- right now working on the SRTIS stuff, but I can add this to my list for the day.

**Comment by jhprinz** [comment]
Thanks. I have enough stuff to do. Need to give a presentation tomorrow in the group seminar...

**Comment by jchodera** [comment]
Your shouldn't be allowed to select bins that clash with the interface
definitions, should you?

**Comment by jhprinz** [comment]
Actually, maybe @dwhswenson can clarify if we actually need the smaller intermediate bins.
I think these do not contribute to the computation of the reweighting factors. it is merely for
plotting and getting the nice probability to reach a certain CV value, is it?

**Comment by dwhswenson** [comment]
You definitely shouldn't do it. "Shouldn't be allowed to" is a more difficult proposition.

The problem is that the volume-based approach to defining interfaces means that the interface boundary is not necessarily unique (there can be an upper bound and a lower bound -- which is the one you're using for your order parameter?) I added `InterfaceSet`s last week, which means that *if* you've assigned a value of the order parameter to each interface, then you can use that information. However, we do not require the user to assign such a value.

We don't have anything that really uses those interface values yet, since I just added the ability last week. We could add something that requires a histogram bin boundary at the interface boundary, if such a boundary is defined.

**Comment by jchodera** [comment]
I'm totally not following. If you already have a decomposition into discrete volumes via an interface set, these are the histogram definitions you should be using, right?

**Comment by dwhswenson** [comment]
I think the main reason to use the denser histogram is to check convergence. It's pretty easy to plot results from individual ensembles on a log scale and eyeball whether you see the same slope. If you don't, you need to sample more.

@bolhuis might have other reasons to do it this way; you'll have to ask him. It may also have an effect on the WHAM calculation: I would need to think more on that. The sum over bins might have some sort of smoothing effect, but I'm not sure.

**Comment by jhprinz** [comment]
This keeps on failing and I am not able to get a travis build pass. Is there a quick fix?

**Comment by jchodera** [comment]
We should really be using the same bin definitions as the interface definitions, and make it *really hard* to do something different---otherwise, we're making it easy for people to screw up.

There are much easier ways to check convergence or error when using the same bin and interface definitions. We can talk about this separately later, but really, we should probably dump WHAM as soon as possible and move to MBAR, which provide uncertainty calculations and a direct assessment of overlap.

**Comment by dwhswenson** [comment]
> This keeps on failing and I am not able to get a travis build pass. Is there a quick fix?

Run the calculations longer.  I think that before, we were somehow not catching a lot of cases that should have raised an error (probably returning rate=0 instead).

**Comment by dwhswenson** [comment]
> We should really be using the same bin definitions as the interface definitions, and make it really hard to do something different

I'm pretty sure @bolhuis will want to be able to easily get well-resolved crossing probability histograms. It's usually one of the first things he wants to see when someone is presenting TIS results. He's on holiday now, but should be included in the conversation before any such changes occur.

**Comment by jchodera** [comment]
I'll open an issue and tag everyone. We shouldn't forget about this, since it could easily cause people to get erroneous results.

**Comment by dwhswenson** [comment]
Can this issue be closed? I think that:

1. The problems with insufficient sampling in the example were fixed by using pre-calculated files for analysis (part of #588).

2. The order parameter was linearized in #552.

3. The other concerns that @jchodera raised regarding selecting how we set up the WHAM histogram bins is in #551. That still hasn't bee implemented, but it is a lot easier now because of #686.

Is anything else that should keep this issue open?
